KCSs/kcs-3.md

```diff
@@ -0,0 +1,310 @@
+
+Protobuf definition:
+
+```proto
+// Arguments
+message total_supply_arguments {}
+// Result
+message total_supply_result {
+   uint64 value = 1 [jstype = JS_STRING];
+}
+```
+
+#### balance_of
+
+Returns how many tokens a specific address holds.
+
+Protobuf definition:
+
+```proto
+// Arguments
+message balance_of_arguments {
+   bytes owner = 1 [(btype) = ADDRESS];
+}
+// Result
+message balance_of_result {
+   uint64 value = 1 [jstype = JS_STRING];
+}
+```
+
+#### get_info (optional)
+
+Returns name, symbol, and decimals in a single call.
+
+Protobuf definition:
+
+```proto
+// Arguments
+message get_info_args {}
+// Result
+message get_info_result {
+   string name = 1;
+   string symbol = 2;
+   uint32 decimals = 3;
+};
+```
+
+#### allowance
+
+Returns the allowance defined for a specific account.
+
+Protobuf definition:
+
+```proto
+// Arguments
+message allowance_args {
+   bytes owner = 1 [(koinos.btype) = ADDRESS];
+   bytes spender = 2 [(koinos.btype) = ADDRESS];
+}
+// Result
+message allowance_result {
+   uint64 value = 1 [jstype = JS_STRING];
+}
+```
+
+#### get_allowances
+
+Returns a paginated list of allowances defined in an account.
+
+Protobuf definition:
+
+```proto
+enum direction {
+   ascending = 0;
+   descending = 1;
+}
+
+message spender_value {
+   bytes spender = 1 [(koinos.btype) = ADDRESS];
+   uint64 value = 2 [jstype = JS_STRING];
+}
+
+// Arguments
+message get_allowances_args {
+   bytes owner = 1 [(koinos.btype) = ADDRESS];
+   bytes start = 2 [(koinos.btype) = ADDRESS];
+   int32 limit = 3;
+   direction direction = 4;
+}
+
+// Result
+message get_allowances_result {
+   bytes owner = 1 [(koinos.btype) = ADDRESS];
+   repeated spender_value allowances = 2;   
+}
+```
+
+### Write methods
+
+#### mint
+
+Used by the contract owner to initially mint the token to a given address.
+
+Protobuf definition:
+
+```proto
+// Arguments
+message mint_arguments {
+   bytes to = 1 [(btype) = ADDRESS];
+   uint64 value = 2 [jstype = JS_STRING];
+}
+// Result
+message mint_result {}
+```
+
+The method should emit a `mint_event` upon success. The event should indicate the recipient of the mint as an impacted account.
+
+```proto
+// Event
+message mint_event {
+   bytes to = 1 [(btype) = ADDRESS];
+   uint64 value = 2 [jstype = JS_STRING];
+}
+```
+
+#### transfer
+
+This will transfer tokens to a new owner. The authorization is checked with the native `check_authority` system call. It is also authorized if the contract of `from` is the one that called the token contract.
+
+Protobuf definition:
+
+```proto
+// Arguments
+message transfer_args {
+   bytes from = 1 [(koinos.btype) = ADDRESS];
+   bytes to = 2 [(koinos.btype) = ADDRESS];
+   uint64 value = 3 [jstype = JS_STRING];
+}
+// Result
+message transfer_result {}
+```
+
+The transfer event should emit a `token_transfer` event upon success. The event should indicate the sender and receiver as impacted accounts.
```

Comment:
> The event should indicate the sender and receiver as impacted accounts.

maybe indicate the order of impacted accounts? first one is receiver, second one is sender.